In-App Links in Channel's Bookmarks

toffguy77 opened the pull request:
# Pull Request: Add In-App Links to Channel Bookmarks

## Summary

Adds support for in-app links (custom URL schemes) in channel bookmarks, allowing users to create bookmarks with mattermost:// URLs that open within the mobile app instead of external browsers.

- Add isValidAppSchemeUrl() and isValidUrlOrAppScheme() utility functions
- Update BookmarkLink component to handle custom app schemes like mattermost://
- Enhance openLink function to properly route app scheme URLs through deep linking
- Update parseDeepLink to handle mattermost:// scheme URLs directly
- Add comprehensive tests for app scheme URL validation
- Update bookmark link description to mention app links support

This enables users to create bookmarks with custom URL schemes that open within the app instead of external browsers, improving user experience for internal navigation.

## Ticket Link

Related to [server-side implementation](https://github.com/mattermost/mattermost/pull/30956) for in-app links feature in channel bookmarks.
Fixes [29145](https://github.com/mattermost/mattermost/issues/29145)

## Checklist

- [x] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ]  Includes text changes and localization file updates

## Release Note

```release-note
Added support for in-app links in channel bookmarks. Users can now create bookmarks with mattermost:// URLs that open within the mobile app.
```

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

toffguy77 commented:
/check-cla

toffguy77 commented on app/utils/url/index.ts:
```diff
@@ -18,6 +18,28 @@ export function isValidUrl(url = '') {
     return regex.test(url);
 }
 
+// Check if the URL has a valid app scheme format
```
Addressed the feedback:
•  isValidAppSchemeUrl now explicitly excludes http/https/ftp/file and validates only custom app schemes (e.g., mattermost://), aligning with the function’s intent and tests.
•  Kept the improved URL normalization (url-parse + stripTrailingSlashes + removeProtocol) and direct handling for mattermost://.
•  Standardized permalink validation to use isValidId.
•  Added handling for Playbook Runs retrospective deep links with a user-facing alert on mobile.
•  Updated tests (jest.mocked usage, removed stray console.log, covered retrospective route).

Let me know if anything else should be adjusted.

matthewbirtch commented:
> @matthewbirtch any thoughts about this feature?

It's a great idea. I don't reall have any issues with supporting this at all as long as it will still work to open these links from the browser?

larkox commented on app/screens/channel_bookmark/components/bookmark_link.tsx:
```diff
@@ -121,7 +129,7 @@ const BookmarkLink = ({disabled, initialUrl = '', resetBookmark, setBookmark}: P
                 <FormattedText
                     id='channel_bookmark_add.link.input.description'
-                    defaultMessage='Add a link to any post, file, or any external link'
+                    defaultMessage='Add a link to any post, file, or any external link. You can also use app links like mattermost://'
```
If we change this, we also have to change the `en.json` file for this to be reflected in the app.

toffguy77 replied:
@larkox Good catch! I've updated the `en.json` file to include the new description text that mentions app links support. The localization string `channel_bookmark_add.link.input.description` now properly reflects the updated functionality.

Note: Since the server-side changes (https://github.com/mattermost/mattermost/pull/30956) are still in Open status and not merged yet, this mobile implementation is ready but will require the server-side support to be fully functional.

larkox replied:
What will happen for old servers (servers without the change)? We need to make sure the app still works (and we don't provide misleading information) if the server doesn't support this.

larkox reviewed (approved):
Codewise looks good to me. A couple minor things to consider, though.

Willyfrog commented:
@esarafianou since this allows for opening links in the app, you might want to take a look too :)

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

toffguy77 commented on app/utils/url/index.ts:
```diff
@@ -18,6 +18,34 @@ export function isValidUrl(url = '') {
+// This supports mattermost:// and other custom app schemes, and explicitly excludes http/https
+export function isValidAppSchemeUrl(url = '') {
+    // Must look like <scheme>://...
+    if (!/^[a-zA-Z][a-zA-Z0-9.+-]*:\/\//.test(url)) {
```
Good point! The validation serves as an early filter to prevent unnecessary parsing attempts for clearly invalid schemes (like http/https/ftp/file). This approach:

1. Avoids parsing overhead for schemes we know we can't handle
2. Provides clearer error handling - we can distinguish between "invalid scheme" vs "valid scheme but parsing failed"
3. Makes the intent explicit in the code flow

However, you're right that we could simplify by letting the parser handle validation. Would you prefer we remove the pre-validation and handle scheme validity within the parsing logic instead?

toffguy77 commented on app/utils/url/links.ts:
```diff
@@ -28,6 +28,25 @@ export const openLink = async (link: string, serverUrl: string, siteURL: string,
 
+    // Handle custom app scheme URLs (like mattermost://)
+    if (isValidAppSchemeUrl(url)) {
+        const match = matchDeepLink(url, serverUrl, siteURL);
```
Done! Removed `isValidAppSchemeUrl` and `isValidUrlOrAppScheme` entirely since `matchDeepLink` now handles both parsing and validation.

toffguy77 commented:
/remove-lifecycle stale

toffguy77 commented:
## Updated to match server-side security validation

Following changes in the server PR https://github.com/mattermost/mattermost/pull/30956, I've updated the mobile client to implement the same URL validation logic.

### Changes:
- Added `isValidLinkURL()` function matching server-side implementation
- Whitelist: `http`, `https`, `mattermost` schemes
- Blacklist: `javascript`, `data`, `file`, `vbscript`, `about` schemes
- Added validation in bookmark creation to prevent XSS attacks

### Why:
The server now validates bookmark URLs using `IsValidLinkURL()` to prevent dangerous schemes like `javascript:` or `data:` that could lead to XSS attacks. The mobile client needs the same validation to:
1. Provide consistent security across all platforms
2. Give immediate feedback to users before server validation
3. Prevent malicious URLs from being processed client-side

All commits have been squashed into a single commit for cleaner history.
